flink-yarn/src/test/java/org/apache/flink/yarn/TestingYarnNodeManagerClientFactory.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.hadoop.yarn.client.api.async.NMClientAsync;
+
+/**
+ * A {@link YarnNodeManagerClientFactory} implementation for testing.
+ */
+public class TestingYarnNodeManagerClientFactory implements YarnNodeManagerClientFactory {
+
+	private NMClientAsync.CallbackHandler callbackHandler;
+	private TestingYarnNMClientAsync testingYarnNMClientAsync;
+	private TestingYarnNMClientAsync.Builder testingYarnNMClientAsyncBuilder = TestingYarnNMClientAsync.builder();
+
+	TestingYarnNMClientAsync.Builder getTestingYarnNMClientAsyncBuilder() {
+		return testingYarnNMClientAsyncBuilder;
+	}
```

Comment:
This ties `TestingYarnNodeManagerClientFactory` with `TestingYarnNMClientAsync`. I think a more flexible approach is to pass in a `Function<NMClientAsync.CallbackHandler, NMClientAsync>`. In that way, we also don't need to provide  `getCallbackHandler()` and `getTestingYarnNMClientAsync()`.

Same for `TestingYarnResourceManagerClientFactory`.